src/libraries/System.Console/src/System/IO/ConsoleStream.cs

```diff
@@ -93,5 +93,37 @@ protected void ValidateWrite(byte[] buffer, int offset, int count)
 
             if (!_canWrite) throw Error.GetWriteNotSupported();
         }
+
+        private static void AccumulateNewLines(StringBuilder accumulator, ReadOnlySpan<char> buffer, Action<string> printer)
+        {
+            int lineStartIndex = 0;
+            for (int i = 0; i < buffer.Length; i++)
```

Comment:
This will always walk the entire buffer.

I think a more performant algorithm would be to walk the buffer backwards, and if you find a newline character, you can print everything up to that character (prepending anything in the buffer), and then put the rest in the buffer. This would be much faster for Console.WriteLine statements, since it would only check the last character in the buffer.

```csharp
for (int i = buffer.Length - 1; i >= 0; i--) {
    if (buffer[i] == '\n') {
        if (accumulator.Length > 0) {
            printer (accumulator + buffer.Slice(0, i + 1));
            accumulator.Clear();
        } else {
            printer (buffer.Slice(0, i + 1));
        }
        if (i < buffer.Length - 1) {
            // there's text after the last newline, add it to the accumulator
            accumulator.Append(buffer.Slice(i + 1));
        }
        return;
    }
}
// no newlines found, add the entire buffer to the accumulator
accumulator.Append(buffer);
```